src/fifelse.c

```diff
@@ -11,127 +11,168 @@ SEXP fifelseR(SEXP l, SEXP a, SEXP b, SEXP na) {
+  const int64_t nmask = len3>1 ? INT64_MAX : 0;
 
   const int *restrict pl = LOGICAL(l);
-  SEXP ans = PROTECT(allocVector(ta, len0)); nprotect++;
-  copyMostAttrib(a, ans);
-
-  bool nonna = !isNull(na);
-  if (nonna) {
-    if (xlength(na) != 1)
-      error(_("Length of 'na' is %"PRId64" but must be 1"), (int64_t)xlength(na));
-    SEXPTYPE tn = TYPEOF(na);
-    if (tn == LGLSXP && LOGICAL(na)[0]==NA_LOGICAL) {
-      nonna = false;
-    } else {
-      if (tn != ta)
-        error(_("'yes' is of type %s but 'na' is of type %s. Please make sure that both arguments have the same type."), type2char(ta), type2char(tn));
-      if (!R_compute_identical(PROTECT(getAttrib(a,R_ClassSymbol)), PROTECT(getAttrib(na,R_ClassSymbol)), 0))
-        error(_("'yes' has different class than 'na'. Please make sure that both arguments have the same class."));
-      UNPROTECT(2);
-      if (isFactor(a)) {
-        if (!R_compute_identical(PROTECT(getAttrib(a,R_LevelsSymbol)), PROTECT(getAttrib(na,R_LevelsSymbol)), 0))
-          error(_("'yes' and 'na' are both type factor but their levels are different."));
-        UNPROTECT(2);
-      }
-    }
-  }
+  SEXP ans = PROTECT(allocVector(tans, len0)); nprotect++;
+  if (!na_a) 
+    copyMostAttrib(a, ans);
+  else if (!na_b)
+    copyMostAttrib(b, ans);
+  else if (!na_n)
+    copyMostAttrib(na, ans);
 
-  switch(ta) {
+  switch(tans) {
   case LGLSXP: {
     int *restrict pans = LOGICAL(ans);
-    const int *restrict pa   = LOGICAL(a);
-    const int *restrict pb   = LOGICAL(b);
-    const int pna = nonna ? LOGICAL(na)[0] : NA_LOGICAL;
+    const int *restrict pa;  if (!na_a) pa  = LOGICAL(a);
+    const int *restrict pb;  if (!na_b) pb  = LOGICAL(b);
+    const int *restrict pna; if (!na_n) pna = LOGICAL(na);
+    const int na = NA_LOGICAL;
     #pragma omp parallel for num_threads(getDTthreads())
     for (int64_t i=0; i<len0; ++i) {
-      pans[i] = pl[i]==0 ? pb[i & bmask] : (pl[i]==1 ? pa[i & amask] : pna);
+      pans[i] = pl[i]==0 ?
+                (na_b ? na : pb[i & bmask]) :
+                pl[i]==1 ? 
+                (na_a ? na : pa[i & amask]) : 
+                (na_n ? na : pna[i & nmask]);
     }
   } break;
   case INTSXP: {
     int *restrict pans = INTEGER(ans);
-    const int *restrict pa   = INTEGER(a);
-    const int *restrict pb   = INTEGER(b);
-    const int pna = nonna ? INTEGER(na)[0] : NA_INTEGER;
+    const int *restrict pa;  if (!na_a) pa  = INTEGER(a);
+    const int *restrict pb;  if (!na_b) pb  = INTEGER(b);
+    const int *restrict pna; if (!na_n) pna = INTEGER(na);
+    const int na = NA_INTEGER;
     #pragma omp parallel for num_threads(getDTthreads())
     for (int64_t i=0; i<len0; ++i) {
-      pans[i] = pl[i]==0 ? pb[i & bmask] : (pl[i]==1 ? pa[i & amask] : pna);
+      pans[i] = pl[i]==0 ?
+                (na_b ? na : pb[i & bmask]) :
+                pl[i]==1 ? 
+                (na_a ? na : pa[i & amask]) : 
+                (na_n ? na : pna[i & nmask]);
     }
```

Comment:
overhead should be measured, most pessimistic case

Reply:
These are pretty fast checks and should be close-to-zero overheads and the benchmark validates this:

### CODE

```r
set.seed(100)
n = 1e6
test_num = runif(n) 
test_true = test = rep(TRUE, n)
test_na = rep(NA, n)
test[test_num <= 1/3] <- FALSE
test[test_num >= 1/3] <- NA
yes = rnorm(n)
no = rnorm(n)

microbenchmark::microbenchmark(
  data.table::fifelse(test_true, yes, no, NA),
  data.table::fifelse(test, yes, no, NA), 
  data.table::fifelse(test_na, yes, no, NA), 
  times = 1e3
)
```

### Before the PR

<img width="415" alt="image" src="https://user-images.githubusercontent.com/8368933/76146717-2812b480-60d0-11ea-8433-81c9fde7432c.png">


### After the PR

<img width="415" alt="image" src="https://user-images.githubusercontent.com/8368933/76146715-2517c400-60d0-11ea-8001-ec7ef6b41464.png">